### PR TITLE
Refactor age calculator spec

### PR DESCRIPTION
**Pull request by jsugarman**
## What

Refactor age calculator spec

[relates to spike](https://dsdmoj.atlassian.net/browse/AP-3629)

For clarity and to cover leap years

## Checklist

Before you ask people to review this PR:

- Tests and rubocop should be passing: `bundle exec rake`
- Github should not be reporting conflicts; you should have recently run `git rebase main`.
- There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- The PR description should say what you changed and why, with a link to the JIRA story.
- You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: spec/services/age_calculator_spec.rb [hunk 1]
```diff
@@ -3,29 +3,51 @@
...
+      it { is_expected.to eq(11) }
+    end
+
+    context "with date of 28th February in a non-leap year" do
+      let(:date) { Date.parse("28-02-2010") }
+
+      it { is_expected.to eq(9) }
+    end
+
+    context "with date of 1st March February in a non-leap year" do
```
Comment (kmahern) [inline spec/services/age_calculator_spec.rb:47]: Should this be "with date of 1st March in a non-leap year"?

**Review by kmahern — APPROVED** [review]
One small typo otherwise lgtm

File: app/services/age_calculator.rb [hunk 2]
```diff
@@ -1,6 +1,6 @@
 class AgeCalculator
-  def self.call(date_of_birth, date)
-    age = date.year - date_of_birth.year
-    date_of_birth > date.years_ago(age) ? age - 1 : age
+  def self.call(date_of_birth, as_of)
+    age = as_of.year - date_of_birth.year
+    date_of_birth > as_of.years_ago(age) ? age - 1 : age
   end
 end
```
Comment (cpjmcquillan) [inline app/services/age_calculator.rb:6]: I appreciate we use the so called "service object" pattern throughout the application, but I wish we could call it like this instead

```ruby
Age.calculate(date_of_birth, as_of)
```

Comment (cpjmcquillan) [reply app/services/age_calculator.rb:6]: Also, would it be clearer if the method was more explicit?

```ruby
class Age
  def self.calculate(date_of_birth, as_of)
    years_since_birth = as_of.year - date_of_birth.year
    next_birthday = date_of_birth.years_since(years_since_birth)
    
    if as_of.before?(next_birthday)
      years_since_birth - 1
    else
      years_since_birth
    end
  end
end
```

**Review by cpjmcquillan — APPROVED** [review]
I added a suggestion that might help reveal the method's intention a bit more clearly.

What do you think?

Other than that, even renaming the variable to `as_of` helps massively, nice work 👍

File: spec/services/age_calculator_spec.rb (hunk 1 above)
Comment (jsugarman) [reply spec/services/age_calculator_spec.rb]: thanks, done

File: app/services/age_calculator.rb (hunk 2 above)
Comment (jsugarman) [reply app/services/age_calculator.rb:6]: I was more interested in refactoring the spec myself and limited myself to only the minor change in the method/class itself.

Feel free to PR one for discussion re: services versus "utility" classes (?!) - i would probably move it out of the services directory in this situation to lib or some such ?!

Comment (jsugarman) [reply app/services/age_calculator.rb:6]: There is also a gem i found btw, not that i would advocate its use, but interesting nonetheless
https://github.com/kenn/age_calculator

**Pull request merged**